### PR TITLE
Cut memory usage

### PR DESCRIPTION
**Pull request by eltang**
Please test these changes before you add them to your pull request. I made them straight on GitHub and they may or may not work.

### CODE REVIEW

**Comment by vifon** [comment]
Thanks.  
Some keys are becoming stuck. For example <kbd>Lower+g</kbd> until I release <kbd>Lower</kbd>. I tried to find the reason but it's pretty hard to read all these bitwise operations at this density.

**Comment by Eric-L-T** [comment]
Sorry, I'm not quite understanding your description of the problem.

**Comment by Eric-L-T** [comment]
What happens if you change all instances of `uint8_t` in `store_or_get_action` to `int8_t`?

**Comment by eltang** [comment]
I found the problems and I'm working on a fix now.

**Comment by eltang** [comment]
Everything should work now. Sorry for wasting your time. This pull request was definitely not ready when I opened it.

File: tmk_core/common/action.c [hunk 1]
```diff
@@ -55,7 +55,7 @@ void action_exec(keyevent_t event)
 
 #if !defined(NO_ACTION_LAYER) && defined(PREVENT_STUCK_MODIFIERS)
 bool disable_action_cache = false;
-int8_t pressed_actions_cache[MATRIX_ROWS][MATRIX_COLS];
+uint8_t source_layers_cache[5][((MATRIX_ROWS * MATRIX_COLS) / 8) ? ((MATRIX_ROWS * MATRIX_COLS) / 8) : 1];
```
Comment (vifon) [inline tmk_core/common/action.c]: I still think it won't work for keyboards with key count not divisible by 8. `(MATRIX_ROWS * MATRIX_COLS) / 8` will round down/truncate the result and we won't have enough array elements for all the keys.

Consider the following modification:

```
uint8_t source_layers_cache[5][((MATRIX_ROWS * MATRIX_COLS) % 8 == 0) ? ((MATRIX_ROWS * MATRIX_COLS) / 8) : ((MATRIX_ROWS * MATRIX_COLS) / 8) + 1];
```

Or even this one:

```
uint8_t source_layers_cache[5][(MATRIX_ROWS * MATRIX_COLS + 7) / 8];
```

Does it make sense?

**Comment by vifon** [comment]
It seems to be working now, thanks. I needed to remove one superfluous parenthesis but otherwise seems good. I'll try it for a day or two.

So the problem was with the signedness of the ints? Subtle... :/

File: tmk_core/common/action.c (hunk 1 above)
Comment (eltang) [inline tmk_core/common/action.c]: Yes. I think the second one is the cleaner solution.

**Comment by eltang** [comment]
I believe that was the problem. Apparently `1 << 7` works correctly only if the right operand is a literal. Otherwise, I think I need to specify `1U`. I just made some other changes for clarity.

**Comment by vifon** [comment]
I was about to merge but since you're still modifying, I'll wait a bit more. Tell me when you're done.

**Comment by eltang** [comment]
You can merge now if you want. I was just doing some refactoring.

**Comment by vifon** [comment]
Merged, thank you.

Sorry for wrapping these lines in such ugly way.

**Comment by eltang** [comment]
I forgot to change the documentation on the memory consumption. Do you want to do that?

**Comment by eltang** [comment]
Also, what do you think about making this feature default now? I have some code that I'm planning to send to Jack that could be simplified if it were.

**Comment by vifon** [comment]
I've just changed the documentation.

Making it default would be nice but it would break too much in my opinion. Either way, it's not for me to decide.

**Comment by eltang** [comment]
Have you noticed any more problems?

**Comment by vifon** [comment]
No issues so far, it's working pretty well.

Like others said the readability/maintainability could be improved. Maybe some helper functions to read and store the encoded data? What do you think?

**Comment by eltang** [comment]
Awesome!

I can see how the code might be hard to read, but what maintenance would have to be done on it unless the maximum number of layers is changed? Are you thinking of using functions like these to encapsulate the messiest parts of the code?

``` C
void store_layer_number(uint8_t key_number, uint8_t layer_number);
uint8_t get_layer_number(uint8_t key_number);
```

Alternatively, would some more drastic changes that create functions like these work?

``` C
void update_source_layers_cache(keypos_t key, uint8_t layer);
uint8_t read_source_layers_cache(keypos_t key);
```

**Comment by vifon** [comment]
Yes, for example like that.

> maintenance would have to be done on it unless the maximum number of layers is changed?

Well, if you could predict the maintenance necessary in the future, there would be no maintenance. ;)

**Comment by eltang** [comment]
Since layers are being stored now, do you think it would make sense to get rid of `store_or_get_action` and move all its code to a function in `action_layer.c`?

**Comment by vifon** [comment]
Hm... Yeah, that would make sense. Will you do it or should I do it?

**Comment by eltang** [comment]
I think it will be more direct for you to do it. It is your pull request, after all. It shouldn't be too difficult, just some simple refactoring.

**Comment by vifon** [comment]
Sure thing. I just wanted to be sure we're not duplicating the effort.

**Comment by vifon** [comment]
Done.

**Comment by eltang** [comment]
I saw what you did and I think the code is much cleaner now, but one of the changes I would have made is to make it so that only layer numbers are passed between `action.c` and `action_layer.c`. The conversion of layer numbers to actions can happen in the former file. What do you think?

**Comment by vifon** [comment]
I think it's fine either way. Some other functions from `action_layer.c` return `action_t` too so it's not really inconsistent.

**Comment by eltang** [comment]
Would you mind if I created a pull request to @tmk with this code? I think I'll make those changes in the process.

**Comment by vifon** [comment]
Go ahead. It's GNU GPL after all.

**Comment by eltang** [comment]
I made a mistake that you can fix. Transpose the indices of the `source_layers_cache` array. The code will run a lot faster that way.

**Comment by vifon** [comment]
Right, the memory cache... I'll do it ASAP.

**Comment by vifon** [comment]
Done. Thanks.

**Pull request merged**